src/fileformats/ocd_georef_fields.cpp

```diff
@@ -0,0 +1,461 @@
+ */
+struct MapperCrsRecord
+{
+	/** CRS id, converted to a string via crs_string_map and passed to
+	 *  CRSTemplateRegistry::find() */
+	const MapperCrs crs_id;
+	/** CRS parameter, converted to a string and passed through to
+	 *  CRSTemplateRegistry::specValues() */
+	const int crs_param;
+};
+
+/**
+ * Plain translation from OCD grid/zone combination into possible Mapper
+ * schemes.
+ *
+ * @see translateMapperToOcd
+ */
+std::vector<MapperCrsRecord> translateOcdToMapper(const int ocd_grid_id,
+                                                  const int ocd_zone_id)
+{
+	std::vector<MapperCrsRecord> retval;
+
+	switch (static_cast<OcdGrid>(ocd_grid_id))
+	{
+	case OcdGrid::LOCAL:
+		retval.push_back({ MapperCrs::LOCAL, 0 });
+		return retval;
+	case OcdGrid::UTM:
+		if (std::abs(ocd_zone_id) >= 1 && std::abs(ocd_zone_id) <= 60)
+		{
+			retval.push_back({ MapperCrs::UTM, ocd_zone_id });
+
+			if (ocd_zone_id > 0)
+				retval.push_back({ MapperCrs::EPSG, 32600 + ocd_zone_id });
+			else
+				retval.push_back({ MapperCrs::EPSG, 32700 - ocd_zone_id });
+		}
+		break;
+
+	case OcdGrid::GERMANY:
+		// DHDN, Gauss-Krüger, Zone 2-5
+		if (ocd_zone_id >= 2 && ocd_zone_id <= 5)
+		{
+			retval.push_back({ MapperCrs::GAUSS_KRUEGER, ocd_zone_id });
+		}
+		// fallthrough - get the EPSG code
+
+	default:
+		for (const auto& record : grid_zone_epsg_codes)
+		{
+			if (static_cast<int>(record.ocd_grid_id) == ocd_grid_id
+			    && record.ocd_zone_id == ocd_zone_id)
+			{
+				retval.push_back({ MapperCrs::EPSG, record.epsg_code });
+				break; // assuming unicity of OCD codes
+			}
+		}
+	}
+
+	return retval;
+}
+
+/**
+ * Imports string 1039 field i.
+ */
+void applyGridAndZone(Georeferencing& georef,
+                      const int combined_ocd_grid_zone,
+                      const std::function<void(const QString&)>& warning_handler)
+{
+	// while OCD codes can be negative, we are keeping grid_id always positive
+	// to match the enums; signum is carried in zone_id
+	const int ocd_grid_id = std::abs(combined_ocd_grid_zone / 1000),
+	          ocd_zone_id = std::abs(combined_ocd_grid_zone % 1000)
+	                        * (combined_ocd_grid_zone < 0 ? -1 : 1);
+
+	// Mapper has multiple ways of expressing some of the CRS's
+	// for now we are working with the first match
+	auto result = translateOcdToMapper(ocd_grid_id, ocd_zone_id);
+
+	if (result.empty())
+	{
+		warning_handler(::OpenOrienteering::OcdFileImport::tr(
+		                    "Could not load the coordinate reference system '%1'.")
+		                .arg(combined_ocd_grid_zone));
+		return;
+	}
+
+	if (result[0].crs_id == MapperCrs::LOCAL)
+		return;
+
+	// look up CRS template by its unique name
+	auto entry_it = std::find_if(std::begin(crs_string_map),
+	                             std::end(crs_string_map),
+	                             [&result](auto e) { return e.key == result[0].crs_id; });
+	if (entry_it == std::end(crs_string_map))
+	{
+		    Q_ASSERT(false);
```

Comment:
`Q_ASSERT(false)` looks like abuse of an assert statement, and in fact it is.
I switched to using `qDebug("Useful message")` instead. It does not abort where the release build wouldn't, it prints a useful message, and it is eliminated in release builds when `QT_NO_DEBUG_OUTPUT` is defined.
(We might even use categorized logging since Qt 5.3, but there are no established logging categories im Mapper.)

For this particular function, an alternative might be a [function-try-block](http://en.cppreference.com/w/cpp/language/function-try-block): Throw an error whenever there is one, and call the warning handler in the `catch` block.

Reply:
Hmmm, `Q_ASSERT()` does not do what I thought it does. Unlike `assert(3)` it does not terminate the program. I'll turn this into `qCritical()` as these two occurences of `Q_ASSERT()` are triggered in case of internal software inconsistency.

Reply:
Even C `assert` won't do anything if NDEBUG is defined (what we do in release builds):

>     If  the  macro  NDEBUG  was  defined  at the moment <assert.h> was last
>     included, the macro assert() generates no code, and hence does  nothing
>     at all.  Otherwise, the macro assert() prints an error message to stan‐
>     ...